### ISSUE TITLE
Search

### ISSUE THREAD

**Issue by nadzree** [Simbul/baker#61]
Was: Does this framework able to do in text search?

It would be great if can search for a related word inside the book. And can we create an index of the topics? Any API for reference?

**Comment by Simbul** [comment]
Searching inside the book is one of the features we'd like to include, but for the time being we want to concentrate on other more pressing issues. You could fork the repository and implement it yourself, anyway, if you feel so inclined.

The index is something that at the moment is better managed by the authors themselves, with an HTML page inside a book.

As for the API, I'm not sure I get what you're referring to. Care to explain? :)

**Comment by nadzree** [comment]
ahh yess..we can put a lineup of titles and link it to the representative pages.. Cool : )
What i mean by API is a short code that can be inserted to ease the development.

We love this framework very much!!

**Comment by MacBoy** [comment]
Just for early discussion, what are your thoughts for doing a search within the book?

What software would do the indexing and searching?

**Comment by nadzree** [comment]
Hi MacBoy.Yes I do meant by searching inside the book content by keyword. But the feature was not available yet. 

Since the framework convert from normal html, an indexing can be done by creating the table of content layout with a link to the page for the content. I never tried with the <h> anchor for auto scrolling to the section. I hope it works. :)

**Comment by jameskatt** [comment]
Yes, I'm interested in the capacity to do search within an eBook. Thanks.

**Comment by jameskatt** [comment]
Before considering the Baker Framework, I considered a similar framework, The TouchBooksReader, framework.  It has a search function and the capacity to download additional books with in-app sales. The only problem is that it costs $500 a book to use, $12,000 for the source code.  Thus, I am grateful for the work of the Baker team.

**Comment by folletto** [comment]
That's even more Adobe's price, around 7000£ (even if it's per year and not the whole source). It also looks like that is more for text ebooks, and not for fully designed books like Baker. :)

Did you try it? :)

**Comment by jameskatt** [comment]
Yes, I did try it.

The author encrypts resources needed to compile and run the book in the iOS simulator.  And he encrypts the resources needed to compile the app for submission.  But he gives the encrypted framework to you free so you can try it out with the iOS simulator.  Then when you want to pay for each book you want to create, you pay him $500 and he gives you the key to compile the final code to send to Apple.  

The author is actually a very nice guy.  It is just too costly to create a series of books with him, as I would like to do.

The Book you create using this framework is any combination of HTML and CSS, just like the Baker Framework. One can add video, graphics, CSS Animations, etc.  When I saw the Baker Framework, my jaw dropped at how similar they are.  Of course, they are for the same goal.  The app store books made with the TouchBooksReader Framework are very very nice, I think.

However, the TouchBooksReader Framework is at a more advanced state than the Baker Framework.  For example.
1. Transition between HTML pages (i.e. chapters of the book) are very very smooth - unlike the jerky, slow transition  of the current Baker Framework.  Thus, TouchBooksReader apps feel like REAL Books, unlike the Baker Framework at this time.
2. Search is enabled for the entire book.
3. One can add Admob ads to the book.
4. One can add in-app purchases of additional books.
5. One is not limited to the 768 px CSS.  You can do sideways scrolling for larger graphics.
6. One set of HTML files can work for both iPad and iPhone in both Portrait and Landscape.
7. etc.

Since the needed resources to compile the app are encrypted rather than open sourced, the TouchBooksReader framework would make me dependent on the author (remaining alive and working on the framework) for updates to keep the app compatible with the rapidly evolving and changing iOS so it doesn't crash on customers and get yanked out of the app store. This is one huge impediment.  

Of course, the cost of the source code puts it out of reach - not only in money but particularly since it forces the user to learn how to code in X-Code in order to keep up with iOS changes - which is the point to avoid by using a framework in the first place.

Thus I hope the Baker Framework can evolve to at least a similar state as the TouchBooksReader Framework.

**Comment by folletto** [comment]
Ok, but the two examples books I've seen don't seem to be able to achieve the same HTML/CSS level Baker does. It seems breaking the content in multiple pages and so on. With content manipulation it's possible to make, for example, the transition smoother. But it's not possible when you build something like Baker _exactly_ like Baker does. :)

To summarize:
- Point 1, is possible at the cost of having something different as far as I know. I may be wrong, but I have to see the code, from the apps it seems that is happening what I'm saying above.
- Points 2 + 4 are in our plan as you can see from the issue tracker and from the contribution of other developers.
- Point 3 is already possible, it's just not out of the box, we don't want to endorse at this point any specific platform.
- Points 5 + 6 are already possible. :)

Is there anywhere I can download the encrypted version to test it?

**Comment by jameskatt** [comment]
You'll have to email the author for the framework.  He'll send it to you.  It will only work on the iOS simulator with your html/css files. It can't be compiled to a final build until you pay him for the serial number. But you will be able to see his code techniques.

**Comment by folletto** [comment]
Ok, I just contacted him. :)

**Comment by designomatt** [comment]
Looks like the $500 version is no longer an option; only $5,000 and $12,000 pricing plans.

**Comment by folletto** [comment]
No answer so far for me.

**Comment by yumyo** [comment]
Just found out the NSHelpManager class which provides parsing, indexing and searching over html files. It belongs to the Apple Help, HTML-based system for providing user assistance. 
We are wondering if it could be the right way to develop the search function in Baker.

**Comment by folletto** [comment]
It might be worth testing. :)

The only thing that I think it could be a blocker is that it seems well tied in into the OS X Help structure, so it might require to conform to the help books format, since there are things in the doc like "For a help book to be searchable, it must have an index file. You can create index files using the Help Indexer utility, described in “Indexing Your Help Book.”", something that we can't do, and looks like only for Mac OS X, not iOS. :)

**Comment by killer1loop** [comment]
This might be of help http://zaldzbugz.posterous.com/how-to-search-a-string-inside-uiwebview

A set of javascript functions to highlight results based on a search string. It uses (stringByEvaluatingJavaScriptFromString) to perform communication between HTML and UI native controls.

It is quite basic as the js functions just provide highlighting of search results, but it might be a good start to work on multiple HTML files and parse results in a list view.

**Comment by folletto** [comment]
Thanks, that's an excellent tutorial. :)

Unfortunately the hard part isn't searching inside a UIWebView document, but looking inside _all_ the documents within a specific book... because these aren't loaded in a UIWebView... :|

**Comment by ghost** [comment]
Not clear on the status of Bobramy's work in issue #369, but LimeMouse apps includes a text search in its "Enhanced Touchbook" The Solar System, which I discovered at http://bakerframework.com/books -- am writing them now for help.

**Comment by swfMonkey** [comment]
@dradal 
Were you able to find out through LimeMouse how they added search? Looking to do the same here but not sure where to add and what directory does the search need to hit exactly.

**Comment by ghost** [comment]
No, LimeMouse did not respond. Work on this issue seems to have ceased, as #369 is closed and contributions to #61 have petered out. Or did I miss something?

**Comment by folletto** [comment]
To be precise, our work on this issue didn't start yet, there were just some contributions from the community but we didn't receive anything workable up until now; #369 was closed because it was a duplicate of this one and well... #61 is this issue. :)

If someone wants to pick up this feature request and implement it, we'd be glad to coordinate. :)

**Comment by kbykhovky** [comment]
I love the baker framework, but also is very interested in the ability to have keyword search in the book.

**Comment by yumyo** [comment]
Swiftype claims to be "the easiest way to add great search to your website or mobile application."
It offers an iOS and Android SDK. While it's a commercial product it comes with a basic free plan too.
Just wondering if it could be an option for Baker.

Some quick links:
https://swiftype.com
https://swiftype.com/documentation/clients/SwiftypeTouch/
https://github.com/swiftype/SwiftypeTouch
https://github.com/swiftype/SwiftypeTouchExample
